### DISCUSSION TITLE
Types from type inference should be shown more simplified

### DISCUSSION THREAD

**Discussion by p-runge** [colinhacks/zod#2358] [Q&A] [answered]
When defining a type with `z.infer` the type does not get recognized as the named, simplified type, but as the more detailed and complex one.
```typescript
const MySchema = z.object({
  foo: z.string(),
});
type MyType = z.infer<typeof MySchema>;

const x: MyType = {
  foo: "bar",
};
```
In this code example hovering `x` doesn't show `const x: MyType` but:
```typescript
const x: {
    foo: string;
}
```
Especially for big and complex types showing `MyType` would drastically improve readability.

**Comment by JacobWeisenburger** [comment, marked as answer]
This is not an issue with Zod, this is a problem with TypeScript Generics.

When you define a type:
```ts
type MyType = { foo: string }
const x: MyType = { foo: 'bar' }
// const x: MyType
```

When you get a type from a generic:
```ts
type GetMyType<T> = T
type MyType = GetMyType<{ foo: string }>
const x: MyType = { foo: 'bar' }
// const x: { foo: string }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by kirill578** [comment]
Anyone has any tips on how to deal with that limitation? it gets tricky when you attempt to look up types of big objects in vscode. [ts-viewer](https://marketplace.visualstudio.com/items?itemName=samxchen.ts-viewer-extension) makes it slightly easier but it is still not the same as having a named type :/
